persist/fs/mmap_linux.go

```diff
@@ -18,58 +18,52 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-package commitlog
+package fs
+
+import "syscall"
 
 const (
-	wordSize    = uint(64)
-	logWordSize = uint(6) // lg(wordSize)
-	wordMask    = wordSize - 1
+	hugePageSize = 2 << 20 // 2mb
 )
 
-type bitSet interface {
-	test(i uint) bool
-	set(i uint)
-	clearAll()
-	getValues() []uint64
-}
-
-type bitSetImpl struct {
-	values []uint64
-}
+func mmap(fd, offset, length int, opts mmapOptions) ([]byte, error) {
+	if length == 0 {
+		// Return an empty slice (but not nil so callers who
+		// use nil to mean something special like not initialized
+		// get back an actual ref)
+		return make([]byte, 0), nil
+	}
 
-func newBitSet(size uint) *bitSetImpl {
-	return &bitSetImpl{values: make([]uint64, bitSetIndexOf(size)+1)}
-}
+	var prot int
+	if opts.read {
+		prot = prot | syscall.PROT_READ
+	}
+	if opts.write {
+		prot = prot | syscall.PROT_WRITE
+	}
 
-func (b *bitSetImpl) test(i uint) bool {
-	idx := bitSetIndexOf(i)
-	if idx >= len(b.values) {
-		return false
+	flags := syscall.MAP_SHARED
+	b, err := syscall.Mmap(fd, int64(offset), length, prot, flags)
+	if err != nil {
+		return nil, err
 	}
-	return b.values[idx]&(1<<(i&wordMask)) != 0
-}
 
-func (b *bitSetImpl) set(i uint) {
-	idx := bitSetIndexOf(i)
-	currLen := len(b.values)
-	if idx >= currLen {
-		newValues := make([]uint64, 2*(idx+1))
-		copy(newValues, b.values)
-		b.values = newValues
+	if length < hugePageSize {
+		return b, nil
 	}
-	b.values[idx] |= 1 << (i & wordMask)
-}
 
-func (b *bitSetImpl) clearAll() {
-	for i := range b.values {
-		b.values[i] = 0
+	if err := syscall.Madvise(b, syscall.MADV_HUGEPAGE); err != nil {
```

Comment:
A comment explaining this would be nice. I found these two links helpful personally: 

1) http://man7.org/linux/man-pages/man2/madvise.2.html
2) https://www.kernel.org/doc/Documentation/vm/transhuge.txt

Seems like the gist of it is that you can reduce TLB misses by making the pages larger, at the risk of wasting a little extra memory. Is that basically it?

Also just out of curiosity, did this make a big difference in your benchmarks or is it just kind of a "best practices" thing?

Reply:
> Seems like the gist of it is that you can reduce TLB misses by making the pages larger, at the risk of wasting a little extra memory. Is that basically it?